azkaban-db/src/main/java/azkaban/db/MySQLDataSource.java

```diff
@@ -58,18 +65,126 @@ public synchronized Connection getConnection() throws SQLException {
+   * than fetch the existing dataSource) if dataSource is broken but not null.
+   */
+  @Override
+  protected synchronized DataSource createDataSource() throws SQLException {
+
+    if (this.dataSource != null) {
+      return this.dataSource;
+    }
+
+    // create factory which returns raw physical connections
+    final ConnectionFactory driverConnectionFactory = createConnectionFactory();
+
+    // Set up the poolable connection factory
+    boolean success = false;
+    final PoolableConnectionFactory poolableConnectionFactory;
+    try {
+      poolableConnectionFactory = createPoolableConnectionFactory(
+          driverConnectionFactory);
+      poolableConnectionFactory.setPoolStatements(isPoolPreparedStatements());
+      poolableConnectionFactory.setMaxOpenPrepatedStatements(getMaxOpenPreparedStatements());
+      success = true;
+    } catch (final SQLException se) {
+      throw se;
+    } catch (final RuntimeException rte) {
+      throw rte;
+    } catch (final Exception ex) {
+      throw new SQLException("Error creating connection factory", ex);
+    }
+
+    if (success) {
+      // create a pool for our connections
+      createConnectionPool(poolableConnectionFactory);
+    }
+
+    // Create the pooling data source to manage connections
+    DataSource newDataSource;
+    success = false;
+    try {
+      newDataSource = createDataSourceInstance();
+      success = true;
+    } catch (final SQLException se) {
+      throw se;
+    } catch (final RuntimeException rte) {
+      throw rte;
+    } catch (final Exception ex) {
+      throw new SQLException("Error creating datasource", ex);
+    } finally {
+      if (!success) {
+        // We should also local connection pool field in this class. In this way, we can manage
+        // connection pool easily.
+        closeConnectionPool(getConnectionPool());
+      }
+    }
+
+    // If initialSize > 0, preload the pool
+    try {
+      for (int i = 0; i < getInitialSize(); i++) {
+        this.getConnectionPool().addObject();
+      }
+    } catch (final Exception e) {
+      closeConnectionPool(getConnectionPool());
+      throw new SQLException("Error preloading the connection pool", e);
+    }
+
+    // If timeBetweenEvictionRunsMillis > 0, start the pool's evictor task
+    startPoolMaintenance();
+
+    this.dataSource = newDataSource;
+    return this.dataSource;
+  }
+
+  private void closeConnectionPool(final GenericObjectPool<PoolableConnection> connectionPool) {
+  }
+
+  private boolean isReadOnly(final Connection conn) throws SQLException {
+    final Statement stmt = conn.createStatement();
+    final ResultSet rs = stmt.executeQuery("SELECT @@global.read_only");
+    if (rs.next()) {
+      final int value = rs.getInt(1);
+      return value != 0;
+    }
+    throw new SQLException("can not fetch read only value from DB");
+  }
+
+  private void sleep(final long milliseconds) {
```

Comment:
consider https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Uninterruptibles.html#sleepUninterruptibly(long, java.util.concurrent.TimeUnit) ?

Reply:
Why?

Reply:
No need to reinvent the wheel

Reply:
Isn't the behavior different? Here we do want the sleep to be interruptable, right?

Reply:
why does it has to be interruptable? maybe we can talk offline.

Reply:
My question is why does it have to be uninterruptable.